User can push to mercurial but email notification not sent - Subject does not have permission for getNotifyConfiguration()

We have a recurrent error that is showing in our logs, which we don't fully understand.

This is the log from Nginx at the time of the failure; the user most likely seems to be using Mercurial Client to push (because of the POST unbundle):

```
THEUSER [27/Apr/2021:09:50:54 +0200] "GET /scm/repo/hg/THEREPOSITORY?cmd=capabilities HTTP/1.1" 200 472 "-" "mercurial/proto-1.0 (Mercurial 5.1)"
THEUSER [27/Apr/2021:09:50:54 +0200] "GET /scm/repo/hg/THEREPOSITORY?cmd=batch HTTP/1.1" 200 7177 "-" "mercurial/proto-1.0 (Mercurial 5.1)"
THEUSER [27/Apr/2021:09:50:54 +0200] "GET /scm/repo/hg/THEREPOSITORY?cmd=listkeys HTTP/1.1" 200 15 "-" "mercurial/proto-1.0 (Mercurial 5.1)"
THEUSER [27/Apr/2021:09:50:55 +0200] "GET /scm/repo/hg/THEREPOSITORY?cmd=listkeys HTTP/1.1" 200 0 "-" "mercurial/proto-1.0 (Mercurial 5.1)"
THEUSER [27/Apr/2021:09:50:55 +0200] "GET /scm/repo/hg/THEREPOSITORY?cmd=branchmap HTTP/1.1" 200 20981 "-" "mercurial/proto-1.0 (Mercurial 5.1)"
THEUSER [27/Apr/2021:09:50:55 +0200] "GET /scm/repo/hg/THEREPOSITORY?cmd=listkeys HTTP/1.1" 200 0 "-" "mercurial/proto-1.0 (Mercurial 5.1)"
THEUSER [27/Apr/2021:09:50:55 +0200] "POST /scm/repo/hg/THEREPOSITORY?cmd=unbundle HTTP/1.1" 200 287 "-" "mercurial/proto-1.0 (Mercurial 5.1)"
THEUSER [27/Apr/2021:09:50:56 +0200] "GET /scm/repo/hg/THEREPOSITORY?cmd=listkeys HTTP/1.1" 200 15 "-" "mercurial/proto-1.0 (Mercurial 5.1)"

```
This is the log from SCM at the same time:

```
2021-04-27 09:50:55.762 [ScmEventBus-1-1] [          ] INFO  sonia.scm.web.security.DefaultAdministrationContext - user THEUSER executes sonia.scm.repository.LastModifiedUpdateListener$LastModifiedPrivileg
edAction as admin
2021-04-27 09:50:55.762 [ScmEventBus-1-1] [          ] INFO  sonia.scm.repository.LastModifiedUpdateListener - update last modified date of repository THEINTERNALID
2021-04-27 09:50:55.762 [ScmEventBus-1-1] [          ] INFO  sonia.scm.repository.DefaultRepositoryManager - modify repository hg/THEREPOSITORY (THEINTERNALID) of type hg
2021-04-27 09:50:56.009 [ScmEventBus-1-0] [          ] ERROR com.github.legman.EventBus - ScmEventBus-1 - could not dispatch event: RepositoryHookEvent{repository=hg/THEREPOSITORY (THEINTERNALID), type=PO
ST_RECEIVE} to handler SynchronizedEventHandler{null, public void sonia.scm.notify.NotifyRepositoryHook.onEvent(sonia.scm.repository.PostReceiveRepositoryHookEvent), true}
java.lang.reflect.InvocationTargetException: null
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at com.github.legman.InvocationContext.invoke(InvocationContext.java:108)
        at com.github.legman.InvocationContext.proceed(InvocationContext.java:101)
        at com.github.legman.micrometer.MicrometerInvocationInterceptor.invoke(MicrometerInvocationInterceptor.java:47)
        at com.github.legman.InvocationContext.proceed(InvocationContext.java:99)
        at com.github.legman.EventHandler.handleEvent(EventHandler.java:103)
        at com.github.legman.SynchronizedEventHandler.handleEvent(SynchronizedEventHandler.java:52)
        at com.github.legman.EventBus.dispatchSynchronous(EventBus.java:452)
        at com.github.legman.EventBus.lambda$dispatch$1(EventBus.java:444)
        at org.apache.shiro.subject.support.SubjectRunnable.doRun(SubjectRunnable.java:120)
        at org.apache.shiro.subject.support.SubjectRunnable.run(SubjectRunnable.java:108)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: org.apache.shiro.authz.UnauthorizedException: Subject does not have permission [repository:notify:THEINTERNALID]
        at org.apache.shiro.authz.ModularRealmAuthorizer.checkPermission(ModularRealmAuthorizer.java:323)
        at org.apache.shiro.mgt.AuthorizingSecurityManager.checkPermission(AuthorizingSecurityManager.java:137)
        at org.apache.shiro.subject.support.DelegatingSubject.checkPermission(DelegatingSubject.java:209)
        at sonia.scm.notify.service.NotifyRepositoryConfigurationService.checkPermission(NotifyRepositoryConfigurationService.java:59)
        at sonia.scm.notify.service.NotifyRepositoryConfigurationService.getNotifyConfiguration(NotifyRepositoryConfigurationService.java:76)
        at sonia.scm.notify.NotifyRepositoryHook.handleEvent(NotifyRepositoryHook.java:98)
        at sonia.scm.notify.NotifyRepositoryHook.handleEvent(NotifyRepositoryHook.java:89)
        at sonia.scm.notify.NotifyRepositoryHook.onEvent(NotifyRepositoryHook.java:76)
        ... 17 common frames omitted
```

Remarks:

- The user THEUSER belongs to a group within THEREPOSITORY which has the default WRITE role. He doesn't have any global permissions.

- The user THEUSER is definitely not an admin (like the first INFO message seems to imply?), and he is not working via browser on the SCM Website at the time.

- The push itself does work - the changeset are pushed into the mercurial repository.

- The repository THEREPOSITORY has Email Notifications set up.

Our current theory is that the email notification does not work if the user doing the push just belongs to WRITE.
Or in other words, that in order for the email notifications to work, the user needs to have some extra permissions.

In my opinion, if a user has the permissions enough to push, then the email notification should be issued - always.

Could you please shed some light here?

SCM-Manager 2.17.1
```
+- scm-activity-plugin@2.0.0
+- scm-git-plugin@2.17.0
+- scm-legacy-plugin@2.17.0
+- scm-mail-plugin@2.5.0
   +- scm-notify-plugin@2.0.1
   +- scm-issuetracker-plugin@3.0.0
+- scm-svn-plugin@2.17.0
+- scm-hg-plugin@2.17.0
+- scm-el-plugin@1.0.0
   +- scm-webhook-plugin@2.2.0
+- scm-webhook-plugin@2.2.0
+- scm-groupmanager-plugin@2.0.0
+- scm-notify-plugin@2.0.1
+- scm-ldap-plugin@2.2.0
+- scm-issuetracker-plugin@3.0.0
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -4,6 +4,10 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+## Unreleased
+### Fixed
+- Wrong permission check ([#11](https://github.com/scm-manager/scm-notify-plugin/pull/11))
+
 ## 2.0.1 - 2020-07-29
 ### Fixed
 - Resolve migration issue when integer fields contain empty strings ([#5](https://github.com/scm-manager/scm-notify-plugin/pull/5))

--- a/src/main/java/sonia/scm/notify/NotifyRepositoryHook.java
+++ b/src/main/java/sonia/scm/notify/NotifyRepositoryHook.java
@@ -95,7 +95,7 @@ public class NotifyRepositoryHook {
   private void handleEvent(Repository repository, Iterable<Changeset> changesets) {
     logger.trace("handle notify event for repository {}", repository.getName());
 
-    NotifyRepositoryConfiguration repositoryConfiguration = configurationService.getNotifyConfiguration(repository.getNamespace(), repository.getName());
+    NotifyRepositoryConfiguration repositoryConfiguration = configurationService.getNotifyConfigurationWithoutPermissionCheck(repository);
 
     if (repositoryConfiguration.isEnabled()) {
       logger.trace("send notification for repository {}", repository.getName());

--- a/src/main/java/sonia/scm/notify/service/NotifyRepositoryConfigurationService.java
+++ b/src/main/java/sonia/scm/notify/service/NotifyRepositoryConfigurationService.java
@@ -74,10 +74,10 @@ public class NotifyRepositoryConfigurationService {
   public NotifyRepositoryConfiguration getNotifyConfiguration(String namespace, String name) {
     Repository repository = getRepository(namespace, name);
     checkPermission(repository);
-    return getNotifyConfiguration(repository);
+    return getNotifyConfigurationWithoutPermissionCheck(repository);
   }
 
-  private NotifyRepositoryConfiguration getNotifyConfiguration(Repository repository) {
+  public NotifyRepositoryConfiguration getNotifyConfigurationWithoutPermissionCheck(Repository repository) {
     ConfigurationStore<NotifyRepositoryConfiguration> store = getStore(repository);
     NotifyRepositoryConfiguration configuration = store.get();
     if (configuration == null) {

